codec/grpc/grpc_codec.go

```diff
@@ -1,33 +1,63 @@
 package grpc
 
-import "fmt"
+import (
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/mem"
+
+	// Guarantee that the built-in proto is called registered before this one
+	// so that it can be replaced.
+	_ "google.golang.org/grpc/encoding/proto"
+)
 
 // Name is the name registered for the proto compressor.
 const Name = "proto"
 
-type Codec struct{}
-
 type vtprotoMessage interface {
-	MarshalVT() ([]byte, error)
+	MarshalToSizedBufferVT(data []byte) (int, error)
 	UnmarshalVT([]byte) error
+	SizeVT() int
+}
+
+type Codec struct {
+	fallback encoding.CodecV2
 }
 
-func (Codec) Marshal(v interface{}) ([]byte, error) {
-	vt, ok := v.(vtprotoMessage)
-	if !ok {
-		return nil, fmt.Errorf("failed to marshal, message is %T (missing vtprotobuf helpers)", v)
+func (Codec) Name() string { return Name }
+
+func (c *Codec) Marshal(v any) (mem.BufferSlice, error) {
+	if m, ok := v.(vtprotoMessage); ok {
+		size := m.SizeVT()
+		if mem.IsBelowBufferPoolingThreshold(size) {
+			buf := make([]byte, 0, size)
```

Comment:
There is a bug on this line IMO and it should go like either `make([]byte, size)` or  `mem.BufferSlice{mem.SliceBuffer(buf[:size])}` on the line 35. Because the original slice reference will have its len still set to 0 effectively returning 0 length byte array from the method.

Reply:
Yes! Apologies -- I have this fixed in my [PR for SpiceDB](https://github.com/authzed/spicedb/pull/2070) but didn't carry over the changes here.